database/migrations/2021_06_09_202500_refactor_photo_model.php

```diff
@@ -0,0 +1,479 @@
+	{
+		Schema::table('photos', function (Blueprint $table) {
+			$table->string('url', 100)->default('');
+			$table->string('thumbUrl', 37)->default('');
+			$table->boolean('thumb2x')->default(false);
+			for ($i = self::VARIANT_ORIGINAL; $i <= self::VARIANT_SMALL; $i++) {
+				$table->integer(self::VARIANT_2_WIDTH_ATTRIBUTE[$i])->unsigned()->nullable()->default(null);
+				$table->integer(self::VARIANT_2_HEIGHT_ATTRIBUTE[$i])->unsigned()->nullable()->default(null);
+			}
+		});
+	}
+
+	/**
+	 * Fills the table size_variants based on the values of the table photos.
+	 *
+	 * The actual work horse of this migration.
+	 */
+	protected function upgradeMigration(): void
+	{
+		DB::beginTransaction();
+
+		$photos = DB::table('photos')->select(array_merge([
+			'id',
+			'url',
+			'type',
+			'thumbUrl',
+			'thumb2x',
+		], self::VARIANT_2_WIDTH_ATTRIBUTE, self::VARIANT_2_HEIGHT_ATTRIBUTE
+		))->lazyById();
+
+		foreach ($photos as $photo) {
+			for ($variant = self::VARIANT_ORIGINAL; $variant <= self::VARIANT_THUMB; $variant++) {
+				if ($this->hasSizeVariant($photo, $variant)) {
+					DB::table('size_variants')->insert([
+						'photo_id' => $photo->id,
+						'size_variant' => $variant,
+						'short_path' => $this->getShortPathOfPhoto($photo, $variant),
+						'width' => $this->getWidth($photo, $variant),
+						'height' => $this->getHeight($photo, $variant),
+					]);
+				}
+			}
+		}
+
+		DB::commit();
+	}
+
+	/**
+	 * Fills the table photos based on the values of the table size_variants.
+	 *
+	 * Reverts {@link RefactorPhotoModel::upgradeMigration()}.
+	 */
+	protected function downgradeMigration(): void
+	{
+		DB::beginTransaction();
+
+		$photos = DB::table('photos')->select([
+			'id',
+			'type',
+			'checksum',
+		])->lazyById();
+
+		foreach ($photos as $photo) {
+			$update = [];
+			$sizeVariants = DB::table('size_variants')
+				->where('photo_id', '=', $photo->id)
+				->get();
+			/** @var object $sizeVariant */
+			foreach ($sizeVariants as $sizeVariant) {
+				$fileExtension = '.' . pathinfo($sizeVariant->short_path, PATHINFO_EXTENSION);
+				$expectedBasename = substr($photo->checksum, 0, 32);
+				if (
+					$sizeVariant->size_variant == self::VARIANT_THUMB2X ||
+					$sizeVariant->size_variant == self::VARIANT_SMALL2X ||
+					$sizeVariant->size_variant == self::VARIANT_MEDIUM2X
+				) {
+					$expectedBasename .= '@2x';
+				}
+				$expectedFilename = $expectedBasename . $fileExtension;
+				$expectedPathPrefix = self::VARIANT_2_PATH_PREFIX[$sizeVariant->size_variant] . '/';
+				if ($sizeVariant->size_variant == self::VARIANT_ORIGINAL && $this->isRaw($photo)) {
+					$expectedPathPrefix = 'raw/';
+				}
+				$expectedShortPath = $expectedPathPrefix . $expectedFilename;
+
+				// Ensure that the size variant is stored at the location which
+				// is expected acc. to the old naming scheme
+				if ($sizeVariant->short_path != $expectedShortPath) {
+					Storage::move($sizeVariant->short_path, $expectedShortPath);
```

Comment:
I believe so that this is the best.

At the moment, there won't be much renaming because the only naming strategy which is implemented is `SizeVariantLegacyNamingStrategy` which exactly mimics today's naming scheme. As long as `SizeVariantLegacyNamingStrategy` will be used no files are moved during downgrade. Only, if `SizeVariantLegacyNamingStrategy` is replaced by some other naming strategy in the future (e.g. a naming strategy which does not use the folder `big/`, `small/`, `thumbs` and `@2` anymore), then files will be moved on downgrade.

This is necessary, because I believe that the downgrade should be such that the state after the downgrade can be used together with the code which has been recent at the time before the migration (see my answer to the previous comment). The current master (`391a1d9`) only stores the `url` and the `thumbUrl` of a photo (actually these URLs are filenames, but that another story). The filenames of all other variants of the photo are derived from these both attributes and the directories `big/ `, `medium/`, `small/` and `thumb/` are prepended as needed. So in order to properly work with the old code, the files must be in those places and their basename must be the same. (This cannot be guaranteed to be the case in the future, if other naming strategies than `SizeVariantLegacyNamingStrategy` will be in effect.)

The alternative would be to simply erase all smaller variants of a photo and then ask the user to re-generated all size variants using the CLI command `lychee:generate_thumbs`. But I thought that would be even more cruel and time consuming.

Reply:
The new code first extracts the base name of the original size variant and only renames the other size variants if necessary. This also solves the issue which you highlighted in the previous remark. Now, if all file names follow the correct pattern and share a common, equal "core basename", then no renaming happens. Also, if this "core basename" does not equal the checksum of the file but is essentially random, no renaming happens. (As long as it is the same for all size variants.)